R/spread_with_duplicates.R

```diff
@@ -0,0 +1,177 @@
+#'
+#' # apply aggregate function
+#' test3 <- data.frame(
+#' col1 = c(1,1,1,1),
+#'   col2  = c("H", "H", "H", "H"),
+#'   key = c("A", "B", "C", "C"),
+#'   value = c(2, 3, 1, 8),
+#'   stringsAsFactors = FALSE
+#' )
+#' spread_with_duplicates(test1, key, value, aggfunc = str_c, collapse = "-")
+#' spread_with_duplicates(test2, key, value, aggfunc = min)
+#' spread_with_duplicates(test2, key, value, aggfunc = mean)
+#'
+#' # same output of spread() if no more than one value per key
+#' library(dplyr)
+#' stocks <- data.frame(
+#'  time = as.Date('2009-01-01') + 0:9,
+#'  X = rnorm(10, 0, 1),
+#'  Y = rnorm(10, 0, 2),
+#'  Z = rnorm(10, 0, 4)
+#' )
+#' stocksm <- stocks %>% gather(stock, price, -time)
+#' stocksm %>% spread_with_duplicates(stock, price)
+#' stocksm %>% spread(stock, price)
+#' stocksm %>% spread_with_duplicates(time, price)
+#' stocksm %>% spread(time, price)
+#'
+#' # Use 'convert = TRUE' to produce variables of mixed type
+#' df <- data.frame(row = rep(c(1, 51), each = 3),
+#'                  var = c("Sepal.Length", "Species", "Species_num"),
+#'                  value = c(5.1, "setosa", 1, 7.0, "versicolor", 2))
+#' df %>% spread_with_duplicates(var, value) %>% str
+#' df %>% spread_with_duplicates(var, value, convert = TRUE) %>% str
+#'
+#' # Use sep non-NULL
+#' spread_with_duplicates(test2, key, value, sep = "_var_")
+#' spread_with_duplicates(df, var, value, sep = "_")
+#' }
+spread_with_duplicates <- function(data, key, value, fill = NA, 
+                                   convert = FALSE, drop = TRUE,
+                                   sep = NULL, aggfunc = NA, ...) {
+  args = list(...)
+  key_var <- vars_pull(names(data), !! enquo(key))
+  value_var <- vars_pull(names(data), !! enquo(value))
+  by = colnames(data)[which(!colnames(data) %in% c(key_var,value_var))]
+  col <- data %>% 
+    pull(key_var) %>% 
+    unique() %>%
+    as.character()
+  
+  data <- map(
+    col, 
+    function(x) data %>% 
+      filter(!! sym(key_var) == x)
+  ) %>%
+    map2(col, ~ change_colname(.x, .y, value_var, key_var)) %>%
+    map2(col, ~ apply_aggfunc(.x, .y,
+                              group_by_col = by,
+                              aggfunc = aggfunc,
+                              args)) %>%
+    map2(col, ~ apply_convert(.x, .y, convert)) %>%
+    map2(col, ~ apply_sep(.x, .y, key_var, sep)) %>%
+    reduce(full_join, by = by)
+  
+  if (!drop) {
+    data <- data %>% complete_(by)
+  }
+  
+  if (!is.na(fill)){
+    key_value_cols <- colnames(data)[which(!colnames(data) %in% by)]
+    data <- data %>% mutate_at(vars(one_of(key_value_cols)),
+                               funs(replace(., is.na(.), fill)))
+  }
+  
+  return(data)
+}
+
+change_colname <- function(data, new_col, value, old_col) {
+  data %>% 
+    rename(!! as.character(new_col) := !!value) %>%
+    select(-one_of(old_col))
+}
+
+
+apply_aggfunc <- function(data, col_name, group_by_col,  aggfunc, args) {
+  if (is.function(aggfunc)) {
+    data %<>%
```

Comment:
I'm just a bit sceptical about adding an additional dependency (`magrittr`) in a package for just this one piping *nice-to feature* ... more dependencies mostly means more trouble. howevr, maybe dplyr ships magrittr anyway?

Reply:
Regular pipes are supported in dplyr, but double ones `%<>%` are not. And I agree: would not include magrittr for this feature only, use `<-` instead.

Reply:
Yes, I see that `%<>%` is used twice. Dependency is not worthy to be included! Solved 9a50683bbda21ce6677f5f4dac8411bdf46095f8